app/Platform/Services/PlayerGameActivityService.php

```diff
@@ -0,0 +1,298 @@
+            ->where('Achievements.GameID', '=', $game->id)
+            ->orderBy('player_achievements.unlocked_at')
+            ->select(['player_achievements.*', 'Achievements.Flags', 'Achievements.Title',
+                      'Achievements.Description', 'Achievements.Points', 'Achievements.BadgeName', 'Achievements.type'])
+            ->get();
+        foreach ($playerAchievements as $playerAchievement) {
+            if ($playerAchievement->unlocked_hardcore_at) {
+                $this->addUnlockEvent($playerAchievement, $playerAchievement->unlocked_hardcore_at, true);
+
+                if ($playerAchievement->unlocked_hardcore_at != $playerAchievement->unlocked_at) {
+                    $this->addUnlockEvent($playerAchievement, $playerAchievement->unlocked_at, false);
+                }
+            } else {
+                $this->addUnlockEvent($playerAchievement, $playerAchievement->unlocked_at, false);
+            }
+
+            $this->achievementsUnlocked++;
+        }
+
+        foreach ($this->sessions as &$session) {
+            $this->sortEvents($session['events']);
+        }
+    }
+
+    private function addUnlockEvent(object $playerAchievement, Carbon $when, bool $hardcore, bool $repeat = false): void
+    {
+        $event = [
+            'type' => 'unlock',
+            'id' => $playerAchievement->achievement_id,
+            'hardcore' => $hardcore,
+            'when' => $when,
+            'achievement' => [ // fields necessary for generating tooltip
+                'ID' => $playerAchievement->achievement_id,
+                'Title' => $playerAchievement->Title,
+                'Description' => $playerAchievement->Description,
+                'Points' => $playerAchievement->Points,
+                'BadgeName' => $playerAchievement->BadgeName,
+                'Flags' => $playerAchievement->Flags,
+                'HardcoreMode' => $hardcore,
+            ],
+        ];
+
+        $unlocker = null;
+        if ($playerAchievement->unlocker_id) {
+            $unlocker = User::firstWhere('id', $playerAchievement->unlocker_id);
+            if ($unlocker) {
+                $event['unlocker'] = $unlocker;
+            }
+        }
+
+        if (!$hardcore && $when < $playerAchievement->unlocked_hardcore_at) {
+            $event['hardcoreLater'] = true;
+        }
+
+        $existingSessionIndex = $this->findSession('player-session', $when);
+        if ($existingSessionIndex < 0) {
+            if ($unlocker) {
+                $existingSessionIndex = $this->generateSession('manual-unlock', $when);
+            } else {
+                $existingSessionIndex = $this->generateSession('generated', $when);
+            }
+        }
+
+        $this->sessions[$existingSessionIndex]['events'][] = $event;
+    }
+
+    private function insertEvent(array &$session, array $event): void
+    {
+        $session['events'][] = $event;
+        $this->sortEvents($session['events']);
```

Comment:
This sort might be redundant, given that `$this->sortEvents($session['events'])` is called at the end of `initialize()`. If it's not redundant, I'm wondering if we can get away with each insertion not triggering a re-sort.

Reply:
This was supposed to insert at the appropriate index to maintain the sort, but it was easier to just append and re-sort. Without that logic, maintaining a sorted list while building it probably isn't necessary. I've removed it for the time-being.